Sources/Datadog/FeaturesIntegration/WebView/WebRUMEventConsumer.swift

```diff
@@ -6,8 +6,91 @@
+        self.webRUMEventMapper = webRUMEventMapper
+        self.contextProvider = contextProvider
+    }
+
+    func consume(event: JSON, eventType: String) throws {
+        let eventData = try JSONSerialization.data(withJSONObject: event, options: [])
+        let jsonDecoder = JSONDecoder()
+        let rumContext = contextProvider?.context
+
+        switch eventType {
+        case "view":
+            let viewEvent = try jsonDecoder.decode(RUMViewEvent.self, from: eventData)
+            let mappedViewEvent = mapIfNeeded(dataModel: viewEvent, context: rumContext, offset: getOffset(viewID: viewEvent.view.id))
+            write(mappedViewEvent)
+        case "action":
+            let actionEvent = try jsonDecoder.decode(RUMViewEvent.self, from: eventData)
+            let mappedActionEvent = mapIfNeeded(dataModel: actionEvent, context: rumContext, offset: getOffset(viewID: actionEvent.view.id))
+            write(mappedActionEvent)
+        case "resource":
+            let resourceEvent = try jsonDecoder.decode(RUMViewEvent.self, from: eventData)
+            let mappedResourceEvent = mapIfNeeded(dataModel: resourceEvent, context: rumContext, offset: getOffset(viewID: resourceEvent.view.id))
+            write(mappedResourceEvent)
+        case "error":
+            let errorEvent = try jsonDecoder.decode(RUMViewEvent.self, from: eventData)
+            let mappedErrorEvent = mapIfNeeded(dataModel: errorEvent, context: rumContext, offset: getOffset(viewID: errorEvent.view.id))
+            write(mappedErrorEvent)
+        case "long_task":
+            let longTaskEvent = try jsonDecoder.decode(RUMViewEvent.self, from: eventData)
+            let mappedLongTaskEvent = mapIfNeeded(dataModel: longTaskEvent, context: rumContext, offset: getOffset(viewID: longTaskEvent.view.id))
+            write(mappedLongTaskEvent)
+        default:
+            userLogger.error("🔥 Web RUM Event Error - Unknown event type: \(eventType)")
+        }
+    }
+
+    private func mapIfNeeded<T: RUMDataModel>(dataModel: T, context: RUMContext?, offset: Offset) -> T {
+        guard let context = context else {
+            return dataModel
+        }
+        // TODO: RUMM-1786 implement mappers
+        let mappedDataModel = dataModel
+        return mappedDataModel
+    }
+
+    private func write<T: RUMDataModel>(_ model: T) {
+        dataWriter?.write(value: model)
+    }
+
+    // MARK: - Time offsets
+
+    // Q: do we really need to cache `offsets`? can't we just read `dateCorrector?.currentCorrection.serverTimeOffset`?
+
+    private typealias Offset = TimeInterval
+    private var offsets = [String: Offset]()
+
+    private func getOffset(viewID: String) -> Offset {
+        var offset = offsets[viewID]
+        if offset == nil {
+            offset = dateCorrector?.currentCorrection.serverTimeOffset
+            offsets[viewID] = offset
+        }
+
+        purgeOffsets()
+        return offset ?? 0.0
+    }
```

Comment:
@mariusc83 is `offsets` an optimization or an obligation? android uses an ordered hashmap data structure which Swift doesn't have; if this is not necessary, i will just read `serverTimeOffset` whenever it's needed instead of caching it. 
is this okay?

Reply:
Native SDK time will always be different than Browser SDK time, hence we need to correct event times using the same source of time. There are two particular cases:
* browser views time must be corrected, to avoid time skew in RUM session timeline;
* browser actions / resources / errors time must be corrected accordingly to the correction in their parent view, to avoid time skew in RUM view timeline.

If we don't correct these times, both RUM session timeline and RUM view timeline in web-ui will be skewed and events will be misaligned. This cannot be ignored, as we work with nanosecond precision.

Reply:
@ncreated the question here is if we need to cache/store `serverTimeOffset` values in `offsets` dictionary?
i suggest, instead of keeping the recently read values in `offsets` dict, reading it directly from `dateCorrector?.currentCorrection.serverTimeOffset` whenever needed.

Reply:
Problem is `dateCorrector?.currentCorrection.serverTimeOffset` might return different offset each time ⚠️. And like I described, we need to apply view's offset to its child events. Hence, we need to cache it somehow 👍. Dictionary is one one way - we could key it by `view.id` which should already be `Hashable`. Question is if/how we will remove things from this cache.

Reply:
conclusion: `offsets` dictionary is essential. resolved.